<a href="https://colab.research.google.com/github/hamidrezanorouzi/numericalMethods/blob/main/Lectures/Lecture07_Numerical_Integration_Differentiation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Numerical Integration and Differentiation**

&nbsp;

Lecturer: **Hamidreza Norouzi**

&nbsp;

### **Notes**❕
This content forms a part of the instructional presentations for the **`numerical methods in chemical engineering`** course designed for undergraduate chemical engineering students at Amirkabir University of Technology.

Feel free to utilize the information and source codes provided in this material, ensuring appropriate acknowledgment of the original document.

The visual elements featured in this document are either original or have been obtained from the following sources, unless specified otherwise:

* Steven C. Chapra, Applied Numerical Methods with Matlab for Engineers and Scientists, 3rd edition, McGraw-Hill (2012).
* Amos Gilat and Vish Subramanian, Numerical Methods for Engineers and Scientists, 3rd edition, Wiley (2014).


<div align="center">
🟪 🟪 🟪
</dive>

---

# 🟢 1) Where do we use them?

# 🔴 2) Integration

## 2-1) Newton-Cotes formulas
* Newton-Cotes formulas are based on the strategy of replacing a complicated function or tabulated data with a polynomial that is easy to integrate:
  $$
  I = \int_a^b{f(x)dx} \thickapprox \int_a^b{f_n(x)dx} \tag{2-1}
  $$
  where the polynomial is:
  $$
  f_n(x) = a_0 + a_1x+a_2x^2+...,a_nx^n \tag{2-2}
  $$

<div align="center">
<img src="https://drive.google.com/uc?id=1i_LQ-H6YAQaYD46ZJ5qA0KQMyKIzUFdh" width="700">
</div>

* How about if we apply the approximation on segments with sub-intervals with smaller than interval [a,b]?

<div align="center">
<img src="https://drive.google.com/uc?id=1CTZJtP00nIn4IvH589DflyNVkF9VyFfw" width="450">
</div>

## 2-2) Trapezoid rule
* It correponds to the case where the ploynomial is first order:

$$
\begin{aligned}
I &= \int_a^b[f(a)+\frac{f(b)-f(a)}{b-a}(x-a)]dx ⇒ \\
I &= (b-a)\frac{f(b)+f(a)}{2}
\end{aligned}
\tag{2-3}
$$

* Geometrically, the trapezoidal rule is equivalent to approximating the area of the trapezoid under the straight line connecting $f(a)$ and $f(b)$:

<div align="center">
<img src="https://drive.google.com/uc?id=12VMro0KWSTEq_7pTW8YqfC5x8kx1xcRO" width="400">
</div>


* The local truncation error of single application of trapezoidal rule is give by:
$$
E_t = \frac{1}{12} f^{"}(ξ)(b-a)^3  \tag{2-4}
$$
  this shows that the error is zero integrating functions up to first-order polynomial.

### 2-2-1) Trapezoidal rule with sub-intervals
* The integration formula can be applied to sub-intervals of domain [a,b] and then these integrals are summed to give the integral over the whole domain.
* Consider that the interval [a,b] is divided into n equal sub-intervals with width $h$:
$$
h = \frac{b-a}{n} \tag{2-5}
$$
Then the intgral over [a,b] can be re-written as:
$$
I = \int_{x_0=a}^{x_n=b} f(x)dx \thickapprox \int_{x_0}^{x_1} f(x)dx + \int_{x_1}^{x_2} f(x)dx + ... + \int_{x_{n-1}}^{x_n} f(x)dx
\tag{2-6}
$$


<div align = "center">
<img src="https://drive.google.com/uc?id=1GgZOKEi-TIi-zrC5amdE609njsBNY_h3" width = "450">
</div>

* Applying trapezoidal rule to each of the above integrals over sub-intervals:
$$
\begin{aligned}
I &= h \frac{f(x_0)+f(x_1)}{2} + h \frac{f(x_1)+f(x_2)}{2} + ...+ h \frac{f(x_{n-1})+f(x_n)}{2} \\
I &= \frac{h}{2}[f(x_0)+ 2 \sum_{i=1}^{n-1}f(x_i) + f(x_n)]
\end{aligned}
\tag{2-7}
$$

* The true error for Eq. (2-7) is:
$$
E_t = \frac{h^3}{12} \sum_{i=1}^{n} f^{"}(ξ)  \tag{2-8}
$$

### ❓ **Example 1:**
Compute the integral of the function $f(x) = 0.2+25x - 200x^2 + 675x^3 - 900x^4 +400x^5$ over interval [0,0.8] using trapezoidal rule with four sub-intervals.

💡 *Solution*

The interval length is:
$$
h = \frac{0.8-0}{4} = 0.2
$$
And the value of function at interval points:
$$
\begin{aligned}
f(0) &= 0.2 \\
f(0.2) &= 1.28800 \\
f(0.4) &= 2.45600 \\
f(0.6) &= 3.46400 \\
f(0.8) &= 0.23200 \\
\end{aligned}
$$

According to Eq. (2-7):
$$
\begin{aligned}
I &= \frac{h}{2}[f(x_0)+ 2 \sum_{i=1}^{3}f(x_i) + f(x_4)]  \\
  &= \frac{0.2}{2}[0.2 + 2(1.28800+2.45600+3.46400)+0.23200] \\
  & = 1.48480
\end{aligned}
$$

The analytical solution of the above integral give the exact value of 1.640533333 for the integral. So, the true relative error is:
$$
ϵ_t = \frac{1.640533333-1.48480}{1.640533333} \times 100 \% = 9.493 \%
$$

### 2-2-2) Code for trapezoidal rule

In [ ]:
# @title Trapezoidal rule integration
def trapz(fx, a, b, n):
  """
  Computes the integral of a function using the trapezoidal rule.

  :param fx: The function to be integrated.
  :param a: The lower limit of integration.
  :param b: The upper limit of integration.
  :param n: The number of subintervals (must be a positive integer).
  :return: The estimated integral value.
  """
  h = (b - a) / n
  integral = 0.5 * (fx(a) + fx(b))

  for i in range(1, n):
    x_i = a + i * h
    integral += fx(x_i)

  integral *= h

  return integral

In [ ]:
# @title Test for trapezoidal rule
exactInt = 1.640533333

fx = lambda x: 0.2+25*x - 200*x**2 + 675*x**3 - 900*x**4 +400*x**5
a = 0
b = 0.8

print("|  n  |     I     |     et    |")
print("|-----|-----------|-----------|")
for n in range(4,81,4):
  I = trapz(fx, a,b,n)
  et = abs(I-exactInt)/exactInt*100
  print(f"| {n:3d} | {I:9.7f} |  {et:6.4f} % |")


|  n  |     I     |     et    |
|-----|-----------|-----------|
|   4 | 1.4848000 |  9.4928 % |
|   8 | 1.6008000 |  2.4220 % |
|  12 | 1.6228082 |  1.0804 % |
|  16 | 1.6305500 |  0.6085 % |
|  20 | 1.6341402 |  0.3897 % |
|  24 | 1.6360922 |  0.2707 % |
|  28 | 1.6372698 |  0.1989 % |
|  32 | 1.6380344 |  0.1523 % |
|  36 | 1.6385587 |  0.1204 % |
|  40 | 1.6389338 |  0.0975 % |
|  44 | 1.6392113 |  0.0806 % |
|  48 | 1.6394224 |  0.0677 % |
|  52 | 1.6395867 |  0.0577 % |
|  56 | 1.6397171 |  0.0498 % |
|  60 | 1.6398223 |  0.0433 % |
|  64 | 1.6399084 |  0.0381 % |
|  68 | 1.6399798 |  0.0337 % |
|  72 | 1.6400395 |  0.0301 % |
|  76 | 1.6400902 |  0.0270 % |
|  80 | 1.6401334 |  0.0244 % |


## 2-3) Simpson’s 1/3 role
* Consider that the polynomial is second-order and this polynomial connects 3 successive points $x_0=a$, $x_1$ and $x_2=b$.
* The integral based on Eq. (2-1) becomes:
$$
\begin{aligned}
I &= \int_{x_0}^{x_2} [\frac{(x-x_1)(x-x_2)}{(x_0-x_1)(x_0-x_2)}f(x_0)+\frac{(x-x_0)(x-x_2)}{(x_1-x_0)(x_1-x_2)}f(x_1)+\frac{(x-x_0)(x-x_1)}{(x_2-x_0)(x_2-x_1)}f(x_2)]dx  \\
⇒ I &= \frac{h}{3}[f(x_0)+4f(x_1)+f(x_2)]
\end{aligned}
\tag{2-9}
$$


<div align="center">
<img src="https://drive.google.com/uc?id=16uMxveXDN792mW-kbYiYnpyaoL7cq8aL" width="400">
</div>

* The local true error associated with this integration (using Eq. (2-9)) is:
$$
E_t = - \frac{1}{90}h^5f^4(\xi)
\tag{2-10}
$$
  which means that this integration method give exact results for polynomials up to 3rd order (zero error).




### 2-3-1) Simpson's 1/3 rule with sub-intervals
* The integration formula can be applied to sub-intervals of domain [a,b] with $n$ sub-intervals, where $n$ is an $even$ number.
* Then the intgral over [a,b] can be re-written as:
$$
I = \int_{x_0=a}^{x_n=b} f(x)dx \thickapprox \int_{x_0}^{x_2} f(x)dx + \int_{x_2}^{x_4} f(x)dx + ... + \int_{x_{n-2}}^{x_n} f(x)dx
\tag{2-11}
$$
  Note that each integral in the above equation covers two successive intervals.

* Applying Simpson's 1/3 formula in Eq. (2-9), we get:

$$
\begin{aligned}
I &= \frac{h}{3}(f(x_0)+4f(x_1)+f(x_2))+\frac{h}{3}(f(x_2)+4f(x_3)+f(x_4))+...+\frac{h}{3}(f(x_{n-2})+4f(x_{n-1})+f(x_n)) \\
⇒ I &= \frac{h}{3}[f(x_0)+ 4 \sum_{i=1,3,5}^{n-1}f(x_i) + 2 \sum_{i=2,4,6}^{n-2}f(x_i)+f(x_n)]
\end{aligned}
\tag{2-12}
$$
<div align="center">
<img src="https://drive.google.com/uc?id=18DhXPlWFzzvkwopSZ82AyfaIdO2pGWSs" width="500">
</div>

### ❓ **Example 2:**
Compute the integral of the function $f(x) = 0.2+25x - 200x^2 + 675x^3 - 900x^4 +400x^5$ over interval [0,0.8] using Simpson's 1/3 rule with four sub-intervals.

💡 *Solution*

The interval length is:
$$
h = \frac{0.8-0}{4} = 0.2
$$
And the value of function at interval points:
$$
\begin{aligned}
f(0) &= 0.2 \\
f(0.2) &= 1.28800 \\
f(0.4) &= 2.45600 \\
f(0.6) &= 3.46400 \\
f(0.8) &= 0.23200 \\
\end{aligned}
$$

According to Eq. (2-12):
$$
\begin{aligned}
I &= \frac{h}{3}[f(x_0)+ 4f(x_1) + 2f(x_2) + 4f(x_3) + f(x_4)]  \\
  &= \frac{0.2}{3}[0.2 + 4(1.28800)+2(2.45600)+4(3.46400)+0.23200] \\
  & = 1.62347
\end{aligned}
$$

The analytical solution of the above integral give the exact value of 1.640533333 for the integral. So, the true relative error is:
$$
ϵ_t = \frac{1.640533333-1.62347}{1.640533333} \times 100 \% = 1.040 \%
$$

### 2-3-2) Code for Simpson's 1/3 rule

In [ ]:
def simpsons(f, a, b, n):

  if n % 2 != 0:
    raise ValueError("Number of subintervals (n) must be even for Simpson's rule."
                     f"Input value is {n}.")

  h = (b - a) / n
  result = f(a) + f(b)

  for i in range(1, n):
    x = a + i * h
    if i % 2 == 0:
      result += 2 * f(x)
    else:
      result += 4 * f(x)

  return result * h/3.0

In [ ]:
exactInt = 1.640533333

fx = lambda x: 0.2+25*x - 200*x**2 + 675*x**3 - 900*x**4 +400*x**5
a = 0
b = 0.8

print("|  n  |     I     |     et    |")
print("|-----|-----------|-----------|")
for n in range(4,81,4):
  I = simpsons(fx, a,b,n)
  et = abs(I-exactInt)/exactInt*100
  print(f"| {n:3d} | {I:9.7f} | {et:7.5f} % |")

|  n  |     I     |     et    |
|-----|-----------|-----------|
|   4 | 1.6234667 | 1.04031 % |
|   8 | 1.6394667 | 0.06502 % |
|  12 | 1.6403226 | 0.01284 % |
|  16 | 1.6404667 | 0.00406 % |
|  20 | 1.6405060 | 0.00166 % |
|  24 | 1.6405202 | 0.00080 % |
|  28 | 1.6405262 | 0.00043 % |
|  32 | 1.6405292 | 0.00025 % |
|  36 | 1.6405307 | 0.00016 % |
|  40 | 1.6405316 | 0.00010 % |
|  44 | 1.6405322 | 0.00007 % |
|  48 | 1.6405325 | 0.00005 % |
|  52 | 1.6405327 | 0.00004 % |
|  56 | 1.6405329 | 0.00003 % |
|  60 | 1.6405330 | 0.00002 % |
|  64 | 1.6405331 | 0.00002 % |
|  68 | 1.6405331 | 0.00001 % |
|  72 | 1.6405332 | 0.00001 % |
|  76 | 1.6405332 | 0.00001 % |
|  80 | 1.6405332 | 0.00001 % |


## 2-4) Simpson’s 3/8 role

* If we use third-order polynomial that connects four successive points $x_0$, $x_1$, $x_2$, and $x_3$, and integrate it over this range $[x_0, x_3]$, we get:
$$
I = \frac{3h}{8}[f(x_0)+3f(x_1)+3f(x_2)+f(x_3)]
\tag{2-13}
$$
* And the local error for this integration is:
$$
E_t = - \frac{3}{80}h^5f^{(4)}(\xi)
\tag{2-14}
$$

* And we want to perform integration over multiple sub-intervals and let's say we number of these sub-intervals is $n=3k$ (multiple of 3):
$$
I = \frac{3h}{8}[f(x_0)+ 3 [\sum_{i=1}^{n/3}f(x_{3i-2})+f(x_{3i-1})]+2\sum_{i=1}^{n/3-1}f(x_{3i}) +f(x_n)]
\tag{2-15}
$$

## 2-5) A note on error of integration methods
*


## 2-6) Integration over intervals with unequal length
* In all above methods, the intervals have equal length.
* In practice, there are many situations where this condition does not hold and we must deal with unequal-sized intervals.
* In this condition, we can apply proper formula (trapezoidal or Simpson) on each of these individual intervals and sum them up.

### ❓ **Example 3:**
Consider the following data from the table below, integrate f(x) from 0 to 0.8. Compare your result with the exact answer 1.6405333.

**a)** Using trapezoidal rule.

**b)** Using a mix of trapezoidal and Simpson's 1/3 rule to achieve minimum error.

| x | f(x) | x| f(x)|
|---|---|---|---|
|0.00| 0.200000| 0.44| 2.842985|
|0.12| 1.309729| 0.54| 3.507297|
|0.22| 1.305241| 0.64| 3.181929|
|0.32| 1.743393| 0.70| 2.363000|
|0.36| 2.074903| 0.80| 0.232000|
|0.40| 2.456000| | |


💡 *Solution*

**Part (a)**

Since all the sub-intervals are not of equal length, we should apply trapezoidal rule on each sub-interval.
$$
I = \frac{0.12}{2}(0.2+1.309729) + \frac{0.1}{2}(1.309729+1.305241)+\frac{0.1}{2}(1.305241+1.743393)+\frac{0.04}{2}(1.743393+2.074903)+\frac{0.04}{2}(2.074903+2.456000)+\frac{0.04}{2}(2.456000+2.842985)+\frac{0.1}{2}(2.842985+3.507297) + \frac{0.1}{2}(3.507297+3.181929)+ \frac{0.06}{2}(3.181929+2.363000) + \frac{0.1}{2}(2.363000+0.232000) = 1.594801
$$

$$
ϵ_t = \begin{vmatrix} \frac{1.6405333-1.594801}{1.6405333}\end{vmatrix} \times 100 \% = 2.79 \%
$$



**Part (b)**

We apply Simpson's 1/3 on two successive equal intervals, 3/8 rules on three successive equal intervals and trapezoidal rule on the rest.

\begin{aligned}
(0.00,0.12) &⇒ \text{trapezoidal} \\
(0.12,0.32) &⇒ \text{Simpson's 1/3} \\
(0.32,0.44) &⇒ \text{Simpson's 3/8} \\
(0.44,0.64) &⇒ \text{Simpson's 1/3} \\
(0.64,0.70) &⇒ \text{trapezoidal} \\
(0.70,0.80) &⇒ \text{trapezoidal} \\
\end{aligned}

&nbsp;


$$
I =  \frac{0.12}{2}(0.2+1.309729) + \frac{0.1}{3}(1.309729+4 \times 1.305241+1.743393) + \frac{3 \times 0.04}{8}(1.743393+3 \times 2.074903+ 3 \times 2.456000+2.842985)+ \frac{0.1}{3}(2.842985+4 \times 3.507297+3.181929) + \frac{0.06}{2}(3.181929+2.363000)+\frac{0.1}{2}(2.363000+0.232000) = 1.603641
$$
$$
ϵ_t = \begin{vmatrix} \frac{1.6405333-1.603641}{1.6405333}\end{vmatrix} \times 100 \% = 2.25 \%
$$

# 🔵 3) Differentiation